src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java

```diff
@@ -284,206 +420,216 @@ private void hoeIfAble(final BlockPos position, final Field field)
-        shouldDumpInventory = true;
-        shouldTryToGetSeed = true;
+        worker.setHeldItem(getHoeSlot());
     }
 
     /**
-     * Terminates the planting process and resets the task.
+     * Handles the offset of the field for the farmer.
      *
-     * @param buildingFarmer the building of the farmer.
-     * @param field          the field being planted.
-     * @return the next state.
+     * @param field the field object.
+     * @return true if successful.
      */
-    @NotNull
-    private AIState terminatePlanting(@NotNull final BuildingFarmer buildingFarmer, @NotNull final Field field)
+    private boolean handleOffset(@NotNull final Field field)
     {
-        if (requestSeeds)
+        if (workingOffset == null)
         {
-            chatSpamFilter.talkWithoutSpam("entity.farmer.NeedSeed", field.getSeed().getItemStackDisplayName(new ItemStack(field.getSeed())));
+            workingOffset = new BlockPos(-field.getLengthMinusX(), 0, -field.getWidthMinusZ());
         }
         else
         {
-            buildingFarmer.getCurrentField().setNeedsWork(false);
+            final int absZ = Math.abs(workingOffset.getZ());
+            if (workingOffset.getZ() >= field.getWidthPlusZ() && workingOffset.getX() >= field.getLengthPlusX())
+            {
+                workingOffset = null;
+                return false;
+            }
+            else if (
+                        (
+                            //If we're checking an even row
+                            ((field.getLengthPlusX() - absZ) % 2 == 0)
+                            && workingOffset.getX() >= field.getLengthPlusX()
+                        )
+                        ||
+                        (
+                            //If we're checking an odd row
+                            ((field.getLengthPlusX() - absZ) % 2 == 1)
+                            && workingOffset.getX() <= -field.getLengthMinusX()
+                        )
+                    )
+            {
+                workingOffset = new BlockPos(workingOffset.getX(), 0, workingOffset.getZ() + 1);
+            }
+            else if ((field.getLengthPlusX() - absZ) % 2 == 0)
+            {
+                workingOffset = new BlockPos(workingOffset.getX() + 1, 0, workingOffset.getZ());
+            }
+            else
+            {
+                workingOffset = new BlockPos(workingOffset.getX() - 1, 0, workingOffset.getZ());
+            }
         }
-
-        resetVariables();
-        return AIState.PREPARING;
+        return true;
     }
 
     /**
      * Handles the offset of the field for the farmer.
+     * Skips to the next harvestable crop, returns true if one was found.
      *
      * @param field the field object.
-     * @return true if successful.
+     * @return true if a harvestable crop was found.
      */
-    private boolean handleOffset(@NotNull final Field field)
+    private boolean handleOffsetHarvest(@NotNull final Field field)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "handleOffsetHarvest" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)